### PR TITLE
fade out sound samples

### LINKED ISSUE #1730: Chex Quest - Clicking Sound Effects
Reported here [https://www.doomworld.com/forum/post/2809339](url)

I can reproduce this clicking and the other observations in that thread on Windows 10. DSITEMUP is a good test candidate (scrolling a submenu.) Somehow, DSDA and PrBoomX don't have this issue. GZDoom and Woof do. Woof even has it back to early releases with SDL Mixer.

Pretty sure this abrupt drop at the end of DSITEMUP is the culprit, and Woof is just playing it faithfully:

dip

Same thing with DSBFG:

Screenshot 2024-06-07 020217

The DSITEMUP that comes with Vanilla Chex 3 doesn't have it, and doesn't pop with a current build of Woof:

https://melodic-spaceship.neocities.org/chex3v/downloads

Screenshot 2024-06-07 015702

This seems like something Woof's fading to avoid clicks could address. But, how do those other ports avoid it? I tried a quick look around and didn't find any obvious "Fix Chex sound" commits.

Edit: Just to help verify things, here's a sound replacement WAD for Chex Quest that allows it to be played in current Woof with no issues that I heard: https://www.doomworld.com/applications/core/interface/file/attachment.php?id=301455

### PR DESCRIPTION
**Pull request by fabiangreffrath**
Fixes #1730

### CODE REVIEW

**Comment by ceski-1** [comment]
Can you add this to the functions in `I_SND_LoadFile` as well?

File: src/i_oalsound.c [hunk 1]
```diff
@@ -610,19 +610,20 @@ static boolean IsPaddedSound(const byte *data, int size)
...
 }
 
-static void FadeInMono8(byte *data, ALsizei size, ALsizei freq)
+static void FadeInOutMono8(byte *data, ALsizei size, ALsizei freq)
 {
     const int fadelen = freq * FADETIME / 1000000;
     int i;
 
-    if (data[0] == 128 || size < fadelen)
+    if ((data[0] == 128 && data[size - 1] == 128) || size < fadelen)
```
Comment (fabiangreffrath) [inline src/i_oalsound.c:618]: I'd prefer to distinguish the "needs fade in" from the "neds fade out" case in this check, but then I'd have to go through the loop over `fadelen` twice in the end.

File: src/i_sndfile.c [hunk 2]
```diff
@@ -521,35 +521,38 @@ static boolean OpenFile(sndfile_t *file, void *data, sf_count_t size)
...
 }
 
-static void FadeInMono16(short *data, ALsizei size, ALsizei freq)
+static void FadeInOutMono16(short *data, ALsizei size, ALsizei freq)
 {
     const int fadelen = freq * FADETIME / 1000000;
     int i;
 
-    if (!data[0] || size / sizeof(short) < fadelen)
+    if ((!data[0] && !data[size - 1]) || size / sizeof(short) < fadelen)
```
Comment (ceski-1) [inline src/i_sndfile.c:529]: I get a "wild pointer" error here and below. Need to set:
```c
const int len = size / sizeof(short);
````
and use
```c
if ((!data[0] && !data[len - 1]) || len < fadelen)
```
and so on.

File: src/i_oalsound.c (hunk 1 above)
Comment (ceski-1) [reply src/i_oalsound.c:618]: I think it's okay to do two loops. It's only looping over 11 or 22 or 44 samples and I'd rather not touch sounds that are already fine at the start or end.

File: src/i_sndfile.c (hunk 2 above)
Comment (ceski-1) [reply src/i_sndfile.c:529]: Maybe this should be passed as an argument for size? I haven't worked it out:

https://github.com/fabiangreffrath/woof/blob/f7b5cd317cf296d600b71acbaa6d134f0aa6fe6a/src/i_sndfile.c#L570

**Comment by ceski-1** [comment]
Test wads: [click_wads.zip](https://github.com/user-attachments/files/15744373/click_wads.zip)

**Comment by fabiangreffrath** [comment]
Thank you!

**Review by ceski-1 — APPROVED** [review]
> Test wads: [click_wads.zip](https://github.com/user-attachments/files/15744373/click_wads.zip)

https://github.com/fabiangreffrath/woof/assets/56656010/67c1e1d8-8579-44e1-be0b-2ac43558e9eb

I think this is good now.

**Review by rfomin — APPROVED** [review]

**Pull request merged**